### PR TITLE
Reimplement Backend.ls()

### LINKED ISSUE #10: Replace _glob() with a simpler function
Currently, a `Backend` class needs to implement a `_glob()` function, which according docstring has to be able to scan into sub-directories and support wildcards. Implementing it for `Artifactory` and `FileSystem` was quite simple, but for other backends we might implement in the future it might become tricky. So I wonder if we should replace it with a simpler function that only returns the file names of a directory (and maybe another one that returns sub-directories). If we want to support recursive search and wildcards, we should implement this in `Backend.glob()` based on this simpler function(s).

### LINKED ISSUE #37: ls() does not raise expected error
```python
import audbackend


backend = audbackend.create(
    'file-system',
    'host',
    'repo',
)
backend.ls('does-not-exist')
```
```
FileNotFoundError: [Errno 2] No such file or directory: '.../host/repo/does-not-exist'
```

and

```python
backend = audbackend.create(
    'artifactory',
    'https://artifactory.audeering.com/artifactory',
    'unittests-public-local',
)
backend.ls('does-not-exist')
```
```
FileNotFoundError: [Errno 2] No such file or directory: '.../unittests-public-local/does-not-exist'
```

But according to the [documentation](https://audeering.github.io/audbackend/api.html#audbackend.Backend.ls) it should raise a `RuntimeError`. However, in all other functions we raise a `FileNotFoundError`, so I would call it a bug and recommend we change the API accordingly.

But there is another problem: currently we let the child implementation raise the error, i.e. we do not have control what error it will actually raise. A proper solution would be to first check if the `path` exists and then call the child class. But here we run into another problem since [exists()](https://audeering.github.io/audbackend/api.html#audbackend.Backend.exists)  only works for files. Instead we could directly call `_exists()` on the child class, which should also work for directories, but the problem is that we need to convert it to backend path first, but [path()](https://audeering.github.io/audbackend/api.html#audbackend.Backend.path) also works only for files.

### LINKED ISSUE #49: Same virtual path can translate into different backend paths
As the following example shows the same virtual path can translate into different backend paths depending on the `ext` argument that we pass:

```python
backend = audbackend.create(
    'artifactory',
    'https://audeering.jfrog.io/artifactory',
    'unittests-public',
)

path = 'folder/file.foo.bar'

for ext in ['', '.foo', '.foo.bar']:
    backend_path = backend._path(
        path,
        '1.0.0',
        ext,
    )
    print(backend_path)
```
```
https://audeering.jfrog.io/artifactory/unittests-public/folder/file.foo.bar/1.0.0/file.foo.bar-1.0.0
https://audeering.jfrog.io/artifactory/unittests-public/folder/file.foo/1.0.0/file.foo-1.0.0.foo
https://audeering.jfrog.io/artifactory/unittests-public/folder/file/1.0.0/file-1.0.0.foo.bar
```

For consistency we have to enforce the same behavior for every backend implementation. So we should at least add a test for it.

### PR DESCRIPTION
**Pull request by frankenjoe**
Relates to #46 

Closes https://github.com/audeering/audbackend/issues/10 #25  #37 https://github.com/audeering/audbackend/issues/48 #49

image

This re-implements `Backend.ls()` by not returning the path on the backend, but a list of virtual paths together with version and extension, so we can close #10 and #25 

In the test we now create three files that have the same virtual path, but we set a different extension. If a backend does not translate it into three different paths, the test will fail, so we can close #49 

We now raise a `FileNotFoundError` error  folder does not exist, so we can close #37

### CODE REVIEW

**Comment by hagenw** [comment]
Would it make sense to provide an option to list only the latest versions of all entries? Or is the use case for `ls()` typically to really access everything?

File: audbackend/core/artifactory.py [hunk 1]
```diff
@@ -109,15 +112,39 @@ def _glob(
 
     def _ls(
             self,
-            path: str,
+            folder: str,
```
Comment (frankenjoe) [inline audbackend/core/artifactory.py:115]: We could set `''` as default value, then it would simply show all files on the backend if no `folder` is specified.

**Comment by hagenw** [comment]
What is the motivation to return an empty list when the virtual path does exists instead of raising a `FileNotFoundError`,
which we also get with:

```python
>>> import os
>>> os.listdir('non-existent')
...
FileNotFoundError: [Errno 2] No such file or directory: 'non-existent'
```

**Comment by frankenjoe** [comment]
> What is the motivation to return an empty list when the virtual path does exists instead

```python
audeer.list_file_names('does-not-exist')
```
```
[]
```

But on the other hand, we also have:

```python
audeer.list_dir_names('does-not-exist')
```
```
FileNotFoundError: [Errno 2] No such file or directory: 'does-not-exist'
```

**Comment by frankenjoe** [comment]
> > Would it make sense to provide an option to list only the latest versions of all entries? Or is the use case for ls() typically to really access everything?
> 
> I also though this might be convenient. So yes let's add it then.

done

**Comment by hagenw** [comment]
Inside `audeer` we don't state that `list_dir_names()` can raise an exception. We should add this or change its behavior. I'm still not decided what would be the best.

File: audbackend/core/artifactory.py (hunk 1 above)
Comment (hagenw) [reply audbackend/core/artifactory.py:115]: OK, then I would set `'/'` as default.

**Comment by frankenjoe** [comment]
> Inside `audeer` we don't state that `list_dir_names()` can raise an exception. We should add this or change its behavior. I'm still not decided what would be the best.

Yes, we should have the same behavior everywhere. For me returning an empty list seems fine, a folder that does not exist also has no content :)

File: audbackend/core/artifactory.py (hunk 1 above)
Comment (frankenjoe) [reply audbackend/core/artifactory.py:115]: done

**Comment by hagenw** [comment]
> > Inside `audeer` we don't state that `list_dir_names()` can raise an exception. We should add this or change its behavior. I'm still not decided what would be the best.
> 
> Yes, we should have the same behavior everywhere. For me returning an empty list seems fine, a folder that does not exist has no content :)

This is true, but it behaves differently then other programs do, e.g.
```bash
$ ls non-existent
ls: cannot access 'non-existent': No such file or directory
```

**Comment by hagenw** [comment]
What is a bit unfortunate now is the following behavior:

```python
>>> backend.ls('folder/name.ext')
...
FileNotFoundError: [Errno 2] No such file or directory: 'folder/name.ext'
```

which is wrong as the file exists. Maybe we should return the single file in this case?

```python
>>> backend.ls('folder/name.ext')
[('folder/name.ext', '1.0.0', '.ext'), ('folder/name.ext', '2.0.0', '.ext')]
```

**Comment by frankenjoe** [comment]
But then we also have to add an `ext` argument since unfortunately the way we implemented the `Artifactory` backend, different extensions can lead to a different paths on the backend, see #49.

**Comment by frankenjoe** [comment]
> What is a bit unfortunate now is the following behavior:
> 
> ```python
> >>> backend.ls('folder/name.ext')
> ...
> FileNotFoundError: [Errno 2] No such file or directory: 'folder/name.ext'
> ```
> 
> which is wrong as the file exists

We now always append `/` to `folder`, so at least the error message would read:

```
FileNotFoundError: [Errno 2] No such file or directory: 'folder/name.ext/'
```

I think this way it's more obvious to the user why it fails.

**Comment by hagenw** [comment]
If we remove `ext` I would again vote to support `backend.ls('folder/name.ext')` as `ls` under Linux does as well.

**Comment by frankenjoe** [comment]
> Would you like to first remove `ext` and then continue here, or would you like to merge this first?

Let's finish this one first.

**Comment by frankenjoe** [comment]
> If we remove `ext` I would again vote to support `backend.ls('folder/name.ext')` as `ls` under Linux does as well.

Agree, could you please create an issue that we don't forget it.

**Comment by hagenw** [comment]
I created https://github.com/audeering/audbackend/issues/59

File: audbackend/core/backend.py [hunk 2]
```diff
@@ -321,28 +321,71 @@ def latest_version(
...
+            folder: str = '/',
+            *,
+            latest_version: bool = False,
+    ) -> typing.List[typing.Tuple[str, str, str]]:
+        r"""List all files under folder.
+
+        Returns a sorted list of tuples
+        with path, extension and version.
+        If file extension is not empty,
+        it starts with a dot.
```
Comment (hagenw) [inline audbackend/core/backend.py:340]: It's not obvious what `it` is referring to. Looking only at the sentence one would guess `sorted list of tuples`, but this makes no sense.

File: audbackend/core/backend.py [hunk 3]
```diff
@@ -321,28 +321,71 @@ def latest_version(
...
+        """  # noqa: E501
+        utils.check_path_for_allowed_chars(folder)
+        if not folder.endswith('/'):
+            folder += '/'
+        paths = self._ls(folder)
+        paths = sorted(paths)
+
+        if len(paths) == 0:
+            if folder == '/':  # pragma: no cover
+                # special case that there are no files on the backend
```
Comment (hagenw) [inline audbackend/core/backend.py:369]: Sounds like we should be able to create a test for it, e.g. creating a new empty repo?

File: audbackend/core/backend.py [hunk 4]
```diff
@@ -321,28 +321,71 @@ def latest_version(
...
+                return []
+            else:
+                utils.raise_file_not_found_error(folder)
+
+        if latest_version:
+            # d[path, ext] = ['1.0.0', '2.0.0']
```
Comment (hagenw) [inline audbackend/core/backend.py:375]:
```suggestion
            # d[(path, ext)] = ['1.0.0', '2.0.0']
```

File: audbackend/core/backend.py [hunk 5]
```diff
@@ -321,28 +321,71 @@ def latest_version(
...
+            d = {}
+            for p, e, v in paths:
+                key = (p, e)
+                if key not in d:
+                    d[key] = []
+                d[key].append(v)
+            # d[path, ext] = '2.0.0'
```
Comment (hagenw) [inline audbackend/core/backend.py:382]:
```suggestion
            # d[(path, ext)] = '2.0.0'
```

File: audbackend/core/backend.py [hunk 6]
```diff
@@ -321,28 +321,71 @@ def latest_version(
...
+            for key, vs in d.items():
+                d[key] = audeer.sort_versions(vs)[-1]
+            # [(path, ext, '2.0.0'), ...]
```
Comment (hagenw) [inline audbackend/core/backend.py:385]: Maybe change to

```suggestion
            # [(path, ext, '2.0.0')]
```
as for me it was not obvious if `...` refer to other versions of the same file or to other (path, ext) combinations.

File: tests/test_backend.py [hunk 7]
```diff
@@ -462,26 +454,62 @@ def test_glob(tmpdir, files, pattern, folder, expected, backend):
...
-    assert backend.ls(backend_path) == expected_content
+    # test
+
+    for folder, expected, expected_latest in [
+        ('/', content, content_latest),
+        ('sub', sub_content, sub_content_latest),
+    ]:
+        folder = backend.join(prefix, folder)
+
+        expected = [  # replace ext where it is None
```
Comment (hagenw) [inline tests/test_backend.py:500]: Maybe

```suggestion
        expected = [  # remove ext where it is None
```

File: tests/test_backend.py [hunk 8]
```diff
@@ -462,26 +454,62 @@ def test_glob(tmpdir, files, pattern, folder, expected, backend):
...
+            (p, f".{p.split('.')[-1]}" if e is None else e, v)
+            for p, e, v in expected
+        ]
+        expected = sorted(expected)
+        assert backend.ls(folder) == expected
+
+        expected_latest = [  # replace ext where it is None
```
Comment (hagenw) [inline tests/test_backend.py:507]: Maybe

```suggestion
        expected_latest = [  # remove ext where it is None
```

File: audbackend/core/artifactory.py [hunk 9]
```diff
@@ -109,15 +112,39 @@ def _glob(
...
-        )
-        return [p.name for p in audfactory.path(path)]
+        r"""List all files under folder."""
+        root = self._folder(folder, '')
+
+        path = audfactory.path(root)
+        try:
+            paths = [str(x) for x in path.glob("**/*") if x.is_file()]
+        except self._non_existing_path_error:  # pragma: nocover
+            paths = []
```
Comment (hagenw) [inline audbackend/core/artifactory.py:124]: What case is this?

If the path does not exist it should raise an error.
If this covers the case of an empty repo with the `/` request, we should add a test for it.

**Comment by hagenw** [comment]
image

Looks like an all-time record ;)

File: audbackend/core/backend.py (hunk 2 above)
Comment (frankenjoe) [reply audbackend/core/backend.py]: It refers to `file extension`, but since we anyway plan to get rid of it in #58 I simply removed the sentence.

File: tests/test_backend.py (hunk 7 above)
Comment (frankenjoe) [reply tests/test_backend.py:500]: If the file extension is `None` we replace it with the default file extension (i.e. characters after last dot), so I think the term "replace" is correct here.

File: tests/test_backend.py (hunk 8 above)
Comment (frankenjoe) [reply tests/test_backend.py:507]: See previous comment.

File: audbackend/core/artifactory.py (hunk 9 above)
Comment (frankenjoe) [reply audbackend/core/artifactory.py:124]: The empty repo case is covered in `Backend.ls()`. So yes, it's the case that the folder does not exist. We then return an empty list and an error is raised in `Backend.ls()`, this way we avoid that the implementation is responsible for raising the expected error. I now added a comment to the docstring that an empty list should be returned by `_ls()` if the folder does not exist.

**Comment by frankenjoe** [comment]
> Looks like an all-time record ;)

Hehehe, it was actually the function to list the files on a backend that kicked-off all those issues we are currently dealing with.

File: audbackend/core/backend.py (hunk 3 above)
Comment (hagenw) [reply audbackend/core/backend.py:371]: Instead of rewirting the tests, you could add an additional test that creates a new empty repo just for this particular test?

File: tests/test_backend.py (hunk 7 above)
Comment (hagenw) [reply tests/test_backend.py:500]: You are right.
I just tought `p.split('.')[-1]` looks not like replacement, but if you read carefully you understand it.

File: audbackend/core/artifactory.py (hunk 9 above)
Comment (hagenw) [reply audbackend/core/artifactory.py:128]: How do you differentiate an empty list returned by a non-existent folder from an empty list returned by an empty folder?

Comment (frankenjoe) [reply audbackend/core/artifactory.py:128]: Since there is no function to create a folder, empty folders cannot exist on the backend. In reality they can of course, but I think it still makes sense to raise always raise an error. Except for the special case of `/`, for which we return an empty list if no file exists yet in the repo.

Comment (frankenjoe) [reply audbackend/core/artifactory.py:128]: So in other words: a folder only exists if it contains at least one file.

Comment (hagenw) [reply audbackend/core/artifactory.py:128]: I'm not that excited about this solution as a user might find other ways to delete all files in a folder, but the folder still exists. I suppose that `backend.exist(folder)` will also return `True` then, and `backend.ls(folder)` will raise an error that it does not exist.

If we want to stay with the behavior that the backend should return a special value we might think of changing the value, e.g. to `None`.

Comment (frankenjoe) [reply audbackend/core/artifactory.py:128]: Another reason for always raising an error if no files are found is that this will guarantee the same behavior on every backend. E.g. assume we put the same file on backend A and B and remove it afterwards. Now assume on backend A removes the folder (or maybe a concept like folders does not even exist), but on backend B we the folder is kept. If we return an empty list if the folder exists, A will now return an empty list, but B raise an error.

File: audbackend/core/backend.py (hunk 3 above)
Comment (frankenjoe) [reply audbackend/core/backend.py:371]: I anyway wanted to rewrite the tests to ensure the backend is cleaned after each test. This is now automatically done by the `backend` fixture. I now added a test for calling `ls()` on an empty repository and removed the pragma comment.

Comment (frankenjoe) [reply audbackend/core/backend.py:371]: I also simplified the tests in a couple of places by using `audeer.touch()` and not encoding the unique ID in the path but already in the repository.

Comment (frankenjoe) [reply audbackend/core/backend.py:371]: And we now keep the list of backends that will be tested in `conftest.py`. So during local debugging the developer can comment out certain backends to speed up the tests.

File: audbackend/core/artifactory.py (hunk 9 above)
Comment (hagenw) [reply audbackend/core/artifactory.py:128]: Yes, makes sense. I forgot that backends don't need support folders, and running `backend.exists()` on a folder would indeed not make any sense.

The only problematic part is that it then still might be useful to stay with our `audfactory` package as this provides an option to the user to check if folders exist on Artifactory and remove them.

Comment (hagenw) [reply audbackend/core/artifactory.py:128]: Is the behavior in `Backend.versions()` that you get `[]` if the folder is empty and a `FileNotFoundError` if the folder does not exist?

Comment (frankenjoe) [reply audbackend/core/artifactory.py:128]: No, currently we never raise an error but return an empty list if the file does not exist.

File: tests/test_backend.py [hunk 10]
```diff
@@ -391,108 +329,95 @@ def test_file(tmpdir, local_file, remote_file, version, ext, backend):
...
-def test_ls(tmpdir, path, content, expected_content, backend):
+def test_ls(tmpdir, backend):
 
-    backend_path = backend.join(
-        pytest.ID,
-        'test_ls',
-        path,
-    )
+    assert backend.ls() == []
+    assert backend.ls('/') == []
```
Comment (hagenw) [inline tests/test_backend.py:370]: This feels still a little bit strange, but I guess `/`, the root of the backend,  is officially not a folder and it makes sense to return an empty list and not raise a `FileNotFoundError`. I think we should add a section to the docstring, stating that this is the only case you can get an empty list?

File: audbackend/core/backend.py [hunk 11]
```diff
@@ -321,28 +321,73 @@ def latest_version(
...
-    ) -> typing.List:  # pragma: no cover
-        r"""List content of path."""
+            folder: str,
+    ) -> typing.List[typing.Tuple[str, str, str]]:  # pragma: no cover
+        r"""List all files under folder.
+
+        Returns an empty list if folder does not exist.
+
+        """
         raise NotImplementedError()
```
Comment (hagenw) [inline audbackend/core/backend.py:331]: We need to add documentation that a developer of a new backend should return an empty list for empty folders (if actual folders exist on the backend). Should we add a comment here or should we add a special developer section to the documentation?

File: tests/test_backend.py (hunk 10 above)
Comment (frankenjoe) [reply tests/test_backend.py:370]: We now say:

image

File: audbackend/core/backend.py (hunk 11 above)
Comment (frankenjoe) [reply audbackend/core/backend.py:331]: I changed the comment to: `Return an empty list if no files match or folder does not exist.` 

And yes, I plan to include a usage section where we discuss how to implement a custom backend.

Comment (hagenw) [reply audbackend/core/backend.py:331]: Good.

I also created https://github.com/audeering/audbackend/issues/60 to track the documentation part.

File: audbackend/core/artifactory.py (hunk 9 above)
Comment (hagenw) [reply audbackend/core/artifactory.py:128]: I created https://github.com/audeering/audbackend/issues/61 to track the proposed change for `backend.versions()`
